frontend-web/webclient/app/Files/LowLevelFileTable.tsx

```diff
@@ -398,6 +420,100 @@ export const LowLevelFileTable: React.FunctionComponent<LowLevelFileTableProps>
+            var products: accounting.ProductNS.Compute[] = []; 
+
+            Object.values(machineSupport.data.productsByProvider).forEach(provider => {
+                const providerProducts = provider.products.filter(product => {
+                    const tool = quickLaunchApp.data!.invocation.tool.tool!;
+                    const backend = tool.description.backend;
+                    switch (backend) {
+                        case "DOCKER":
+                            return product.support.docker.enabled;
+                        case "SINGULARITY":
+                            return false;
+                        case "VIRTUAL_MACHINE":
+                            return product.support.virtualMachine.enabled &&
+                                (tool.description.supportedProviders ?? [])
+                                    .some(p => p === product.product.category.provider);
+                    }
+                })
+                .filter(product =>
+                    wallet.data.items.some(wallet => productCategoryEquals(product.product.category, wallet.category))
+                )
+                .map(it => it.product)
+                providerProducts.forEach(it => products.push(it));
+            })
+
+            setAvailableProducts(products);
+        }
+    }, [quickLaunchApp, machineSupport]);
+
+    React.useEffect(() => {
+        if (quickLaunchApp.data !== null && props.path) {
+            if (availableProducts.length === 1) {
+                quickLaunchJob(
+                    quickLaunchApp.data,
+                    {
+                        id: availableProducts[0].id,
+                        category: availableProducts[0].category.id,
+                        provider: availableProducts[0].category.provider
+                    },
+                    props.path,
+                    history
+                )
+            } else if (availableProducts.length > 1) {
+                setSelectProduct(true);
+            }
+        }
+    }, [availableProducts]);
+
+    const onQuickLaunch = React.useCallback((app: QuickLaunchApp) => {
+        fetchWallet(accounting.products.browse({filterUsable: true, filterArea: "COMPUTE", itemsPerPage: 250, includeBalance: true}));
+        setSelectedQuickLaunchApp(app);
+    }, []);
+
+    const quickLaunchOnSelectedMachine = React.useCallback(() => {
+        if (selectedMachine !== null && quickLaunchApp.data !== null && props.path) {
+            quickLaunchJob(
+                quickLaunchApp.data,
+                {
+                    id: selectedMachine.id,
+                    category: selectedMachine.category.id,
+                    provider: selectedMachine.category.provider
+                },
+                props.path,
+                history
+            )
+        } else {
+            snackbarStore.addFailure("Please select a machine type", true)
+        }
+    }, [selectedMachine, quickLaunchApp, props.path]);
```

Comment:
Have you tested these, make sure the closures update correctly?

Reply:
I have double checked now, they all seem to update properly.